src/keyutil/keyutil.c

```diff
@@ -277,11 +303,124 @@ static int verb_public(int argc, char *argv[], void *userdata) {
+
+        int x509_mdnid = 0, x509_pknid = 0;
+        if (X509_get_signature_info(certificate, &x509_mdnid, &x509_pknid, /* secbits= */ NULL, /* flags= */ NULL) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to get X.509 digest NID/PK: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (X509_ALGOR_set0(signer_info->digest_alg, OBJ_nid2obj(x509_mdnid), V_ASN1_NULL, /* pval= */ NULL) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set digest alg: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (X509_ALGOR_set0(signer_info->digest_enc_alg, OBJ_nid2obj(x509_pknid), V_ASN1_NULL, /* pval= */ NULL) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set digest enc alg: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        /* Create new PKCS7 using X509 certificate */
+
+        _cleanup_(PKCS7_freep) PKCS7 *pkcs7 = PKCS7_new();
+        if (!pkcs7)
+                return log_oom();
+
+        if (PKCS7_set_type(pkcs7, NID_pkcs7_signed) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS#7 type: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_content_new(pkcs7, NID_pkcs7_data) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS#7 content: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_set_detached(pkcs7, true) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS#7 detached attribute: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        if (PKCS7_add_certificate(pkcs7, certificate) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS#7 certificate: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
+        /* Add PKCS1 signature to PKCS7_SIGNER_INFO */
+
+        ASN1_STRING_set0(signer_info->enc_digest, TAKE_PTR(pkcs1), pkcs1_len);
+
+        /* Add PKCS7_SIGNER_INFO to PKCS7 */
+
+        if (PKCS7_add_signer(pkcs7, signer_info) == 0)
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to set PKCS#7 signer info: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+        TAKE_PTR(signer_info);
+
+        _cleanup_fclose_ FILE *output = fopen(arg_output, "we");
+        if (!output)
+                return log_error_errno(errno, "Could not open PKCS#7 output file %s: %m", arg_output);
+
+        if (!i2d_PKCS7_fp(output, pkcs7))
+                return log_error_errno(SYNTHETIC_ERRNO(EIO), "Failed to write PKCS#7 file: %s",
+                                       ERR_error_string(ERR_get_error(), NULL));
+
```

Comment:
this should really fflush() and check for errors. i.e. we should fail the tool if for some reason we cannot actually write the file we are supposed to generate. fink_tmpfile() will do that for you (among other things)